src/src/com/microsoft/aad/adal/TokenCacheAccessor.java

```diff
@@ -0,0 +1,317 @@
+            removeTokenCacheItem(cachedItem, resource);
+        }
+    }
+    
+    /**
+     * Update token cache with returned auth result. 
+     * @param result
+     */
+    void updateTokenCache(final String resource, final String clientId, final AuthenticationResult result) {
+        if (result == null || StringExtensions.IsNullOrBlank(result.getAccessToken())) {
+            return;
+        }
+        
+        if (result.getUserInfo() != null) {
+            // update cache entry with displayableId
+            if (!StringExtensions.IsNullOrBlank(result.getUserInfo().getDisplayableId())) {
+                setItemToCacheForUser(resource, clientId, result, result.getUserInfo().getDisplayableId());
+            }
+            
+            // update cache entry with userId
+            if (!StringExtensions.IsNullOrBlank(result.getUserInfo().getUserId())) {
+                setItemToCacheForUser(resource, clientId, result, result.getUserInfo().getUserId());
+            }
+        }
+        
+        // update for empty userid
+        setItemToCacheForUser(resource, clientId, result, null);
+    }
+    
+    /**
+     * Remove token from cache.
+     * {@link RefreshItem#mKeysWithUser} is holding a list of keys related to user for removal. 
+     * 1) If refresh with resource specific token cache entry, clear RT with key(R,C,U,A)
+     * 2) If refresh with MRRT, clear RT (C,U,A) and (R,C,U,A)
+     * 3) if refresh with FRT, clear RT with (U,A) 
+     * @throws AuthenticationException 
+     */
+    void removeTokenCacheItem(final TokenCacheItem tokenCacheItem, final String resource) 
+            throws AuthenticationException {
+        final List<String> keys;
+        final TokenEntryType tokenEntryType = tokenCacheItem.getTokenEntryType();
+        switch (tokenEntryType) {
+        case REGULAR_TOKEN_ENTRY : 
+            Logger.v(TAG, "Regular RT was used to get access token, remove entries "
+                    + "for regular RT entries.");
+            keys = getKeyListToRemoveForRT(tokenCacheItem);
+            break;
+        case MRRT_TOKEN_ENTRY : 
+            Logger.v(TAG, "MRRT was used to get access token, remove entries for both "
+                    + "MRRT entries and regular RT entries.");
+            keys = getKeyListToRemoveForMRRT(tokenCacheItem, resource);
+            
+            final TokenCacheItem regularRTItem = new TokenCacheItem(tokenCacheItem);
+            regularRTItem.setResource(resource);
+            keys.addAll(getKeyListToRemoveForRT(regularRTItem));
+            break;
+        case FRT_TOKEN_ENTRY : 
+            Logger.v(TAG, "FRT was used to get access token, remove entries for "
+                    + "FRT entries.");
+            keys = getKeyListToRemoveForFRT(tokenCacheItem);
+            break;
+        default : 
+            throw new AuthenticationException(ADALError.INVALID_TOKEN_CACHE_ITEM);
```

Comment:
I don't want to throw runtime in this case, what's the problem of throwing a checked exception in this case?

---
In reply to: [63403737](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/611#discussion_r63403737) [](ancestors = 63403737)